src/vsr/routing.zig

```diff
@@ -0,0 +1,877 @@
+//! while the primary tries an alternative route. The primary captures replication latency for a
+//! route (that is, the arrival time of prepare_ok messages). If the latency for an alternative
+//! route is sufficiently better than current latency, the route is switched. Note that latency
+//! includes both network and disk latency.
+//!
+//! The experiment schedule is defined randomly. All replicas share the same RNG seed, so no
+//! coordination is needed to launch an experiment!
+//!
+//! To remove outliers, the experiment take two ops. op and op|1 are either both experimental, or
+//! both non-experimental. Experiment's cost is the average of two costs. The cost for a route
+//! is exponential moving average.
+const std = @import("std");
+const assert = std.debug.assert;
+const maybe = stdx.maybe;
+const constants = @import("../constants.zig");
+const stdx = @import("../stdx.zig");
+const Ratio = stdx.PRNG.Ratio;
+const ratio = stdx.PRNG.ratio;
+const Instant = stdx.Instant;
+const Duration = stdx.Duration;
+
+const history_max = constants.pipeline_prepare_queue_max * 2;
+
+// This constant serves two purposes:
+// - First, as a cap for all latencies, to make sure that u64 computations can't overflow.
+// - Second, to simplify cost function computations, replica we haven't heard from has this latency.
+const latency_max: Duration = .{ .ns = 10 * std.time.ns_per_hour };
+
+replica: u8,
+replica_count: u8,
+standby_count: u8,
+view: u32,
+
+/// The route we currently use.
+a: Route,
+a_cost: ?Cost, // Computed as exponential weighted moving average.
+
+/// The best alternative, which might be better or worse than a.
+b: Route,
+b_cost: ?Cost, // Computed as an average of two experiments.
+
+experiment_chance: Ratio = ratio(1, 20),
+
+history: [history_max]OpHistory,
+
+pub const Route = stdx.BoundedArrayType(u8, constants.replicas_max);
+const OpHistory = struct {
+    op: u64,
+    prepare: Instant,
+    prepare_ok: [constants.replicas_max]Duration = .{latency_max} ** constants.replicas_max,
+    present: stdx.BitSetType(constants.replicas_max) = .{},
+
+    fn root() OpHistory {
+        return .{
+            .op = 0,
+            .prepare = .{ .ns = 0 },
+        };
+    }
+
+    fn record_prepare_ok(
+        op_history: *OpHistory,
+        replica: u8,
+        now: Instant,
+    ) void {
+        const latency = now.duration_since(op_history.prepare);
+        if (op_history.present.is_set(replica)) {
+            assert(op_history.prepare_ok[replica].ns <= latency.ns);
+            return;
+        }
+        assert(!op_history.present.is_set(replica));
+        assert(op_history.prepare_ok[replica].ns == latency_max.ns);
+
+        op_history.prepare_ok[replica] = Duration.min(latency, latency_max);
+        op_history.present.set(replica);
+    }
+};
+
+const Cost = struct {
+    // Left-biased median latency matches replication quorum latency, and directly contributes to
+    // user-visible latency.
+    median: Duration,
```

Comment:
Originally I parametrized the whole thing by the quorum size, but I think hard-coding median is the same in effect and leads to a very nice definition of cost.